CodingStandard.md

```diff
@@ -0,0 +1,1425 @@
+  device-independent interface to hardware services.
+
+**Object-Oriented Software Design Guidelines**
+
+- Where appropriate, make use of proven design patterns. If a design pattern
+  adds complexity, make sure the benefit outweighs the added complexity.
+- Use of interfaces as a means to decrease coupling is recommended; allow
+  clients to depend on interfaces to software elements, not on their
+  implementations.
+- Composition/aggregation is recommended instead of implementation inheritance.
+- Use of inheritance solely to provision for unlikely future product
+  improvements should be avoided. Abstracting an interface and separating it
+  from an implementation as reasonable accommodation for change may be
+  appropriate.
+- Avoid allocating too many responsibilities to a single class.
+
+
+Additional Guidelines and Conventions
+================================================================================
+
+This section contains recommendations, not project standards.
+
+The guidelines should be considered during code review and deviations in the
+code base should be justified or resolved as soon as possible.
+
+**File Naming**
+
+C++ code should adhere to these naming conventions:
+
+  - All source code filenames should be lower case.
+
+  - C++ files should use the .cpp and .h extensions (.hpp extension
+    should not be used).
+
+  - Header files that are #included in the body of another file (e.g. to define
+    a reusable lookup table) should use the -inline.h suffix.
+
+  - All C++ files associated with a design package should be prefixed with the
+    package namespace name followed by an underscore. For example if `nav` is a
+    namespace representing the Navigation package, then
+    `nav_localize.h` and `nav_localize.cpp` are files belonging to the Navigation 
+
+package.
+
+  - Files should be named with the primary entity they contain. Example:
+    `nav_localize.h` contains declaration of class `nav::Localize`.
+    Exceptions are allowed when a file contains an assortment of different
+    types or entities.
+
+  - Files defining an abstract base class should end in `_interface.h`, rather
+    than begin with `i`.
+
+
+**Identifier Naming**
+
+C++ code should adhere to the naming conventions below to promote consistency
+throughout the project.
+```
+  - Pre-Processor Macros: `UPPER_CASE`
+  - Class, Struct, Union, and Enum Types: `CamelCase`
+  - Functions and Class Methods: `lower_case`
+  - Compile-time constants (const, constexpr, enumerators, etc.): `kCamelCase`
+  - Variables and Function Parameters: `lower_case`
+  - constexpr Methods and Functions: `lower_case`
+  - Non-const struct member variables: `lower_case`
+  - Non-const class member variables: `lower_case_` (incl. trailing undercore)
+  - Const member variables: `kCamelCase`
+  - Template type parameters: `CamelCase`
+  - Template value parameters: `kCamelCase`
+  - Namespaces (4 characters max) : `lower_case`
+
```

Comment:
Pretty sure we should find a beautifier or static analysis tool to automate this.

Reply:
`clang-tidy`